### PR TITLE
Make sure the resource_partitioner is properly destructed on hpx::finalize

### LINKED ISSUE #2918: hpx::init calls wrong (first) callback when called multiple times
The following code does not work as expected:

```
#include <hpx/hpx_init.hpp>
#include <iostream>

int hpx_init_test(std::string s, int argc, char ** argv)
{
    std::cout << "string is " << s << std::endl;
    return hpx::finalize();
}

int main (int argc, char *argv[])
{
    using hpx::util::placeholders::_1;
    using hpx::util::placeholders::_2;

    hpx::util::function_nonser<int(int, char**)> callback1 = hpx::util::bind(&hpx_init_test, "first", _1, _2);
    hpx::init(callback1, "name", argc, argv, hpx::runtime_mode_default);

    hpx::util::function_nonser<int(int, char**)> callback2 = hpx::util::bind(&hpx_init_test, "second", _1, _2);
    hpx::init(callback2, "name", argc, argv, hpx::runtime_mode_default);

    hpx::util::function_nonser<int(int, char**)> callback3 = hpx::util::bind(&hpx_init_test, "third", _1, _2);
    hpx::init(callback3, "name", argc, argv, hpx::runtime_mode_default);

    return 0;
}
```

Expected output:
```
first
second
third
```

Actual output:
```
first
first
first
```

This is on latest master (`f032046`).

### PR DESCRIPTION
**Pull request by hkaiser**
- this fixes #2918

### CODE REVIEW

**Comment by hkaiser** [comment]
@msimberg please verify whether this solves the issue you reported

File: tests/regressions/multiple_init_2918.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
+//  Copyright (c) 20017 msimberg
```
Comment (msimberg) [inline tests/regressions/multiple_init_2918.cpp]: This should probably be 2017 Mikael Simberg.

**Comment by msimberg** [comment]
@hkaiser Yes, thank you, this does solve the issue.

**Comment by biddisco** [comment]
This failed on circle CI - but only because the hpx:bind=non was fixed on another branch and not included in this build. I think it's safe to merge.

**Pull request merged**